### ISSUE TITLE
Support http.HTTPStatus

### ISSUE THREAD

**Issue by floqqi** [falconry/falcon#1135]
It would be great to support [`http.HTTPStatus`](https://docs.python.org/3/library/http.html#http.HTTPStatus) from the standard library (3.5+).

Could possibly look like:

```python
from http import HTTPStatus

class MyResource:
    def on_get(self, req, res):
        res.status = HTTPStatus.NOT_FOUND  # instead of falcon.HTTP_404
```

**Comment by kgriffs** [comment]
Not a bad idea. Since 1.4 is on the home stretch, let's look at this for 2.0. For the sake of backwards-compatibility, we will probably want to allow both the enum and the raw string.

**Comment by vytas7** [comment]
As discussed on Gitter, I can take a stab at this.

Basic design:
* Python 3.5+ ``http.HTTPStatus`` or its subclass in Falcon will be THE way to set status
* Python 3.5+ ``http.HTTPStatus`` will be backported to be used in older Python versions
* Falcon 1.x status constants ``falcon.HTTP_*`` will be aliased to the new enum values to maintain backward compatibility for the most common usecases

Note that, strictly speaking, it will still be a breaking change in type compatibility, so setting status to raw strings not obtained from ``falcon.HTTP_*`` constants will yield a type error.

**Comment by vytas7** [comment]
@kgriffs on Gitter, we agreed that a lookup table to speed things up is a reasonable approach to proceed.

You also indirectly implied it could be a status property:

> Kurt Griffiths
> Agreed, it should be good enough. One area where Falcon is opinionated is encouraging apps to follow the RFCs. Inventing app-specific codes sort of violates the uniform interface.
> In a pinch, I suppose someone could subclass Request (probably *Response* // Vytas) and override the status prop

Another alternative I had in my mind was having the lookup table on the API side.
The drawback is that should one desire to use some completely nonstandard type (yes, we are going to maintain 7XX, teapots and coffee machines), it will be even harder and dirtier to hook that in. Unless we add an API for that.

Thoughts?

In the attached ``test.py``, the lookup table on the API (no use of properties) seems to outdo even the current approach (at least on ``py27``) due to the absence ``if six.PY2`` clause.

**Comment by xgfone** [comment]
A good idea!

When using falcon, I like `int` as the status. But falcon does not support it, so I wrap it.
```python
import falcon

STATUS_CODES = {}
for v in vars(falcon.status_codes).values():
    if isinstance(v, str):
        try:
            code = int(v.split(maxsplit=1)[0])
            STATUS_CODES[code] = v
        except Exception:
            pass

class Resource(object):
    def status(self, code):
        return STATUS_CODES[code]

    def send_response(self, resp, status=200, body=None, content_type=None):
        resp.status = self.status(status)
        if content_type:
            resp.content_type = content_type
        if body:
            resp.body = body

   def on_get(self, req, resp):
        self.send_response(resp)
```

But, it's tedious. If falcon supports it natively, it's perfect!

**Comment by vytas7** [comment]
Resuming work on this!

**Comment by kgriffs** [comment]
An alternative approach to this that I think has worked out pretty well on the ASGI side:

Use the new `falcon.util.misc.http_status_to_code()` method from #1573 to convert the value so that it can be used in the response. It uses `functools.lru_cache` to make it blazing fast.

We'll also need to update the docs to let users know that they can now set `resp.status` to anything that `http_status_to_code()` accepts (we may want to update the tutorials/examples as well to use instances of http.HTTPStatus from the standard library).

**Comment by vytas7** [comment]
I'm assigning myself for the time being, if anyone wants to help out, feel free to volunteer!
Otherwise I'll try to squeeze this in for the 3.0 release myself.

**Comment by vytas7** [comment]
@kgriffs have you read upon [`functools.lru_cache` thread safety issues](https://bugs.python.org/issue28969) btw?

As I understand the problem is fixed on the newest versions of CPython, and probably backported even to down to 3.5? Might it still be an issue if popular distributions still ship the unpatched variant?

**Comment by kgriffs** [comment]
Ugh, that's awful. Based on some quick tests, it looks like it is fixed as of 3.5.4+, 3.6.1+, 3.7.0+ and 3.8.0. I suppose we could detect the version and then swap out the decorator for a NOP function as needed?
